### ISSUE TITLE
GLSL: RelaxedPrecision decoration applied to operation result IDs is not correctly translated to essl

### ISSUE THREAD

**Issue by Dredhog** [KhronosGroup/SPIRV-Cross#1868]
Hi, It appears that SPIRV-Cross might not be translating `RelaxedPrecision` decorations correctly into essl.

This is similar to case https://github.com/KhronosGroup/SPIRV-Cross/issues/1855 but more general.

Here's an example HLSL fragment shader:
```
min16float RelaxedCalc(min16float3 input)
{
    min16float temp0 = input.x;
    min16float temp1 = input.y;
    min16float temp2 = input.z;
    return temp0 * temp1 * temp2;
}

float FloatCalc(min16float3 input)
{
    float temp0 = input.x;
    float temp1 = input.y;
    float temp2 = input.z;
    return temp0 * temp1 * temp2;
}

float4 PSMain(min16float4 color : COLOR) : SV_TARGET
{
    return float4(RelaxedCalc(color), FloatCalc(color), 0, 0);
}
```

Here's the SPIR-V that DXC generates for it:
```
; SPIR-V
; Version: 1.0
; Generator: Google spiregg; 0
; Bound: 21
; Schema: 0
               OpCapability Shader
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %PSMain "PSMain" %in_var_COLOR %out_var_SV_TARGET
               OpExecutionMode %PSMain OriginUpperLeft
               OpSource HLSL 600
               OpName %in_var_COLOR "in.var.COLOR"
               OpName %out_var_SV_TARGET "out.var.SV_TARGET"
               OpName %PSMain "PSMain"
               OpDecorate %in_var_COLOR Location 0
               OpDecorate %out_var_SV_TARGET Location 0
               OpDecorate %in_var_COLOR RelaxedPrecision
               OpDecorate %4 RelaxedPrecision
               OpDecorate %5 RelaxedPrecision
               OpDecorate %6 RelaxedPrecision
               OpDecorate %7 RelaxedPrecision
               OpDecorate %8 RelaxedPrecision
               OpDecorate %9 RelaxedPrecision
      %float = OpTypeFloat 32
    %float_0 = OpConstant %float 0
    %v4float = OpTypeVector %float 4
%_ptr_Input_v4float = OpTypePointer Input %v4float
%_ptr_Output_v4float = OpTypePointer Output %v4float
       %void = OpTypeVoid
         %16 = OpTypeFunction %void
%in_var_COLOR = OpVariable %_ptr_Input_v4float Input
%out_var_SV_TARGET = OpVariable %_ptr_Output_v4float Output
     %PSMain = OpFunction %void None %16
         %17 = OpLabel
          %4 = OpLoad %v4float %in_var_COLOR
          %5 = OpCompositeExtract %float %4 0
          %6 = OpCompositeExtract %float %4 1
          %7 = OpCompositeExtract %float %4 2
          %8 = OpFMul %float %5 %6
          %9 = OpFMul %float %8 %7
         %18 = OpFMul %float %5 %6
         %19 = OpFMul %float %18 %7
         %20 = OpCompositeConstruct %v4float %9 %19 %float_0 %float_0
               OpStore %out_var_SV_TARGET %20
               OpReturn
               OpFunctionEnd
```
SPIRV-Cross generates the following essl:
```
#version 320 es
precision mediump float;
precision highp int;

layout(location = 0) in vec4 in_var_COLOR;
layout(location = 0) out highp vec4 out_var_SV_TARGET;

void main()
{
    out_var_SV_TARGET = vec4((in_var_COLOR.x * in_var_COLOR.y) * in_var_COLOR.z, (in_var_COLOR.x * in_var_COLOR.y) * in_var_COLOR.z, 0.0, 0.0);
}
```

Basically, in the `essl` produced by SPIRV-Cross, all of the multiplications are allowed to run at `mediump`, even though only the result IDs of the multiplications in `RelaxedCalc` i.e. `%8`, `%9` are decorated with `RelaxedPrecision`, and the result IDs corresponding to `FloatCalc` i.e. `%19`, `%20` don't have `RelaxedPrecision` decorations, which to my understanding is not correct.

What I would expect is that whenever an operation doesn't have a `RelaxedPrecision` decoration the `mediump` inputs for it in the `essl` should first be moved into `highp` temporaries (only necessary when there isn't at least one `highp` input already). The opposite should also be true i.e. whenever an operation is decorated with `RelaxedPrecision` all of the operands which aren't already `mediump` should be first moved into `mediump` temporaries.

Here's my understanding of the situation:
* In SPIR-V decorating the result of e.g. an `OpFMul` with `RelaxedPrecision` means it should execute at relaxed precision.
* Not decorating  `OpFMul` with `RelaxedPrecision` means it should execute at full precision, this is independent of if the operands have `RelaxedPrecision` decorations.
* For `RelaxedPrecision` variables the only operations which implicitly inherit the decoration seem to be loads and stores, not arithmetic ops, this is how I understand the [SPIR-V spec](https://www.khronos.org/registry/SPIR-V/specs/unified1/SPIRV.html#_relaxed_precision).
* On the other hand, in essl, an operation's precision is independent of the destination's precision but depends on the precision of the operands (I believe there are some rare exceptions when none of the operands have a precision specified, but I don't think this applies to SPIRV-Cross as it always provides a default precision).

**Comment by Dredhog** [comment]
cc @AlbinBernhardssonARM

**Comment by HansKristian-Work** [comment]
Supporting this strictly requires a rewrite of how temporary expressions are emitted. :\

**Comment by HansKristian-Work** [comment]
The most annoying part of this is that an SPIR-V ID can be used in multiple precision contexts. A variable can be used in a relaxed precision expression and a highp expression.

**Comment by Dredhog** [comment]
> We might have to introduce highp T spvHighpConstant(T), highp T spvHighpCast(mediump T) and mediump T spvMediumpCast(highp T). That should force precision inline.

This seems mostly reasonable, though I'm somewhat concerned that drivers would fail to perform optimizations as well due to the extra indirection compared to adding some extra temporaries .. :\

Also does the argument type being `mediump`/`highp` matter, would the following two functions not be sufficient? `highp T spvHighpCast(T) { return T; }`, `mediump T spvMediumpCast(T) { return T; }`

> The other option is to analyze expressions which are highp and promote its expression dependency chain to highp somehow. So many edge cases though ... :\

Hmm, curious, I'm most likely thinking overly naively, but since all the operations that are supposed to run at `RelaxedPrecision` are directly marked as such do you need to analyse anything except the immediate arguments coming into them? What are some edge cases?

**Comment by HansKristian-Work** [comment]
> What are some edge cases?

Anything related to loading variables and/or textures can inherit precision in SPIR-V. There's also stores to mediump variables that allows something implicit to happen.

**Comment by Dredhog** [comment]
>Anything related to loading variables and/or textures can inherit precision in SPIR-V. There's also stores to mediump variables that allows something implicit to happen.

**Comment by Dredhog** [comment]
> Anything related to loading variables and/or textures can inherit precision in SPIR-V. There's also stores to mediump variables that allows something implicit to happen.

Hmm, reading the spec multiple times it does sound like you're right about Load inheritance, but it's written very vaguely. I don't see anything about textures or stores though...

With respect o `opLoad` inheritance:

`When applied to a variable, function parameter, or structure member, all loads and stores from the decorated object may be treated as though they were decorated with RelaxedPrecision` and `However, a load or store might reduce the precision (as allowed by RelaxedPrecision) of the destination value.`

it does sound like the `OpDecorate %4 RelaxedPrecision` in the original example is implicit could be removed:
```
          OpDecorate %4 RelaxedPrecision
          ...
          %4 = OpLoad %v4float %in_var_COLOR
```
and `%4` would in theory still be interpreted as `RelaxedPrecision` (if the driver implements the spec correctly)...
So to handle all SPIR-V that might be required, but so far it seems that at least for DXC's output `opLoad` IDs like `%4` are marked explicitly.

Also, I don't see anything written about `OpCompositeExtract` inheriting precision, so I guess `%5`, `%6`, `%7` would never inherit the precision and still need explicit decorations?

>Adding extra temporaries is not a practical solution I think. It complicates expressions since it'd have to select between two temporary variants, and we cannot easily generate temporaries inside loop continue statements, so I'd really like to avoid that path if possible.

In general, we'd prefer the codegen with minimal casts, i.e. in terms of codegen simplicity & perf predictability it would be preferrable to have temporaries whenever possible and casts for the exceptions like the loop continue case you described, but if casting is the best path you see forward it's still better than having the current precision bugs we're seeing as those cause visual regressions, not just performance issues.

**Comment by Dredhog** [comment]
Hmm, I guess since SPIRV-Cross might inline `OpCompositeExtract` you do need to follow the chain to know what precision the original value is so you don't e.g. inline it thinking it will be full precision when the value it's reading from is `RelaxedPrecision`...

**Comment by HansKristian-Work** [comment]
I'll give this a deeper look today.

**Comment by Dredhog** [comment]
Thank you!
An FYI in case this helps, I've prototyped a workaround for this bug by adding temporaries for the operands whose precision doesn't math the result ID precision to the SPIR-V using OpCopyObject (I could share the code it if you are interested)

At least for the handful of shaders I've tested this appears to force SPIR-V to produce essl that matches the SPIR-V in terms of precision (though I have not yet done extensive testing)

Also, if you intend to do precision analysis with the Mali Offline Compiler, I would recommend using version 7.5.0 or above as 7.4 and below seem to have bugs about how they calculate fp16 use for ESSL shaders

**Comment by Dredhog** [comment]
Ok, thank you!

I had a quick look and I've noticed that my initial example does not yet work correctly (SPIRV-Cross does moves the inputs into temporaries, but then uses them in both contexts):

i.e. SPIRV-Cross generates this:
```
#version 300 es
precision mediump float;
precision highp int;

in vec4 in_var_COLOR;
layout(location = 0) out highp vec4 out_var_SV_TARGET;

void main()
{
    highp float _5 = in_var_COLOR.x;
    highp float _6 = in_var_COLOR.y;
    out_var_SV_TARGET = vec4((_5 * _6) * in_var_COLOR.z, (_5 * _6) * in_var_COLOR.z, 0.0, 0.0);
}
```

**Comment by HansKristian-Work** [comment]
What do you mean "correctly"? Promoting to highp is still correct.

**Comment by HansKristian-Work** [comment]
To clarify, "need", or "should"? It's legal for RelaxedPrecision to run at high precision, just maybe not optimal.

Being able to deal with a temporary in two precision contexts will not be trivial, but there might be some hacks that can be done with OpCopyObject and rewrite code to insert those automatically as needed ...

**Comment by Dredhog** [comment]
Ok, so I wasn't aware of the scale of the intervention you were planning on doing. I agree that the `RelaxedPrecision` decorations are optional optimization hints, and ignoring or "promoting" the inputs is technically legal.

However, my expectation for "correctness" was that SPIRV-Cross "should" preserve the precision intent of the original SPIR-V as we were seeing noticeable reductions in fp16 use during the `SPIR-V` to `essl` translation (most likely since operation precision in most cases is implicitly promoted to `highp` instead of demoted and inlining compounds this effect)

**Comment by Dredhog** [comment]
I guess if SIPRV-Cross won't prevent us from implementing the fixup using `OpCopyObject` on the application side then this change is fine. However, it would definitely be nice if the generated `essl` matched `SPIR-V`without needing workarounds :).

I'll need to do more testing though (I'm already seeing a weird demotion of an `OpCopyObject` temporary with this PR version of SPIRV-Cross, it does not change the end result, but am not yet sure why it happens at all)

**Comment by Dredhog** [comment]
Here's a slightly different SPIR-V example (I've also added `OpCopyObject` temporaries which start at ID `1000000` to fix the precision codegen)
```
; SPIR-V
; Version: 1.0
; Generator: Google spiregg; 0
; Bound: 1000005
; Schema: 0
               OpCapability Shader
               OpExtension "SPV_GOOGLE_hlsl_functionality1"
               OpExtension "SPV_GOOGLE_user_type"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %frag "frag" %gl_FragCoord %out_var_SV_Target0
               OpExecutionMode %frag OriginUpperLeft
               OpSource HLSL 600
               OpName %out_var_SV_Target0 "out.var.SV_Target0"
               OpName %frag "frag"
               OpDecorate %gl_FragCoord BuiltIn FragCoord
               OpDecorateString %gl_FragCoord UserSemantic "SV_POSITION"
               OpDecorateString %out_var_SV_Target0 UserSemantic "SV_Target0"
               OpDecorate %out_var_SV_Target0 Location 0
               OpDecorate %4 RelaxedPrecision
               OpDecorate %5 RelaxedPrecision
               OpDecorate %6 RelaxedPrecision
               OpDecorate %7 RelaxedPrecision
               OpDecorate %8 RelaxedPrecision
               OpDecorate %1000000 RelaxedPrecision
               OpDecorate %1000001 RelaxedPrecision
      %float = OpTypeFloat 32
    %float_1 = OpConstant %float 1
    %v3float = OpTypeVector %float 3
    %float_0 = OpConstant %float 0
    %v4float = OpTypeVector %float 4
%_ptr_Input_v4float = OpTypePointer Input %v4float
%_ptr_Output_v4float = OpTypePointer Output %v4float
       %void = OpTypeVoid
         %17 = OpTypeFunction %void
%gl_FragCoord = OpVariable %_ptr_Input_v4float Input
%out_var_SV_Target0 = OpVariable %_ptr_Output_v4float Output
%float_1_00986195 = OpConstant %float 1.00986195
         %19 = OpConstantComposite %v3float %float_1_00986195 %float_1_00986195 %float_1_00986195
       %frag = OpFunction %void None %17
         %20 = OpLabel
         %21 = OpLoad %v4float %gl_FragCoord
         %22 = OpVectorShuffle %v3float %21 %21 0 1 2
    %1000000 = OpCopyObject %v3float %22
    %1000001 = OpCopyObject %v3float %19
          %4 = OpFMul %v3float %1000000 %1000001
          %5 = OpCompositeExtract %float %4 0
          %6 = OpCompositeExtract %float %4 1
          %7 = OpFMul %float %5 %6
          %8 = OpFMul %float %7 %6
    %1000002 = OpCopyObject %float %5
    %1000003 = OpCopyObject %float %6
         %23 = OpFMul %float %1000002 %1000003
    %1000004 = OpCopyObject %float %6
         %24 = OpFMul %float %23 %1000004
         %25 = OpCompositeConstruct %v4float %8 %24 %float_0 %float_1
               OpStore %out_var_SV_Target0 %25
               OpReturn
               OpFunctionEnd
```

an earlier version of SPIRV-Cross `(Git commit: 0d4ce028 Timestamp: 2022-04-15T10:34:50)` generates this:
```
#version 300 es
precision mediump float;
precision highp int;

layout(location = 0) out highp vec4 out_var_SV_Target0;

void main()
{
    vec3 _1000000 = gl_FragCoord.xyz;
    vec3 _1000001 = vec3(1.00986194610595703125);
    vec3 _4 = _1000000 * _1000001;
    float _5 = _4.x;
    float _6 = _4.y;
    highp float _1000002 = _5;
    highp float _1000003 = _6;
    highp float _1000004 = _6;
    out_var_SV_Target0 = vec4((_5 * _6) * _6, (_1000002 * _1000003) * _1000004, 0.0, 1.0);
}
```

and this PR version generates this:
```
#version 300 es
precision mediump float;
precision highp int;

layout(location = 0) out highp vec4 out_var_SV_Target0;

void main()
{
    vec3 _1000000 = gl_FragCoord.xyz;
    vec3 _1000001 = vec3(1.00986194610595703125);
    vec3 _4 = _1000000 * _1000001;
    float _5 = _4.x;
    float _6 = _4.y;
    highp float _1000002 = _5;
    highp float _1000003 = _6;
    float _1000004 = _6;
    out_var_SV_Target0 = vec4((_5 * _6) * _6, (_1000002 * _1000003) * _1000004, 0.0, 1.0);
}
```

So here `%1000004` is the result of `OpCopyObject`, it is not decoration as `RelaxedPrecision` and is used in an `OpFMul` which is also not `RelaxedPrecision`, but SPIRV-Cross for some reason demotes it from `highp float` to `float`.

Just sharing as it looks like it might be a bug in this PR (possibly not necessarily specific to `OpCopyObject`)

**Comment by HansKristian-Work** [comment]
OpCopyObject will inherit the precision of the input because lack of RelaxedPrecision in SPIR-V does not imply promotion to highp until it's consumed. I'll see what can be done about injecting magic double temporaries, but this is straddling deep into heroics territory.

**Comment by HansKristian-Work** [comment]
Went deeper this time. It handles alias to mediump as well this time.

**Comment by HansKristian-Work** [comment]
```
; SPIR-V
; Version: 1.0
; Generator: Google spiregg; 0
; Bound: 1000005
; Schema: 0
               OpCapability Shader
               OpExtension "SPV_GOOGLE_hlsl_functionality1"
               OpExtension "SPV_GOOGLE_user_type"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %frag "frag" %gl_FragCoord %out_var_SV_Target0
               OpExecutionMode %frag OriginUpperLeft
               OpSource HLSL 600
               OpName %out_var_SV_Target0 "out.var.SV_Target0"
               OpName %frag "frag"
               OpDecorate %gl_FragCoord BuiltIn FragCoord
               OpDecorateString %gl_FragCoord UserSemantic "SV_POSITION"
               OpDecorateString %out_var_SV_Target0 UserSemantic "SV_Target0"
               OpDecorate %out_var_SV_Target0 Location 0
               OpDecorate %4 RelaxedPrecision
               OpDecorate %5 RelaxedPrecision
               OpDecorate %6 RelaxedPrecision
               OpDecorate %7 RelaxedPrecision
               OpDecorate %8 RelaxedPrecision
      %float = OpTypeFloat 32
    %float_1 = OpConstant %float 1
    %v3float = OpTypeVector %float 3
    %float_0 = OpConstant %float 0
    %v4float = OpTypeVector %float 4
%_ptr_Input_v4float = OpTypePointer Input %v4float
%_ptr_Output_v4float = OpTypePointer Output %v4float
       %void = OpTypeVoid
         %17 = OpTypeFunction %void
%gl_FragCoord = OpVariable %_ptr_Input_v4float Input
%out_var_SV_Target0 = OpVariable %_ptr_Output_v4float Output
%float_1_00986195 = OpConstant %float 1.00986195
         %19 = OpConstantComposite %v3float %float_1_00986195 %float_1_00986195 %float_1_00986195
       %frag = OpFunction %void None %17
         %20 = OpLabel
         %21 = OpLoad %v4float %gl_FragCoord
         %22 = OpVectorShuffle %v3float %21 %21 0 1 2

          %4 = OpFMul %v3float %22 %19
          %5 = OpCompositeExtract %float %4 0
          %6 = OpCompositeExtract %float %4 1
          %7 = OpFMul %float %5 %6
          %8 = OpFMul %float %7 %6

         %23 = OpFMul %float %5 %6

         %24 = OpFMul %float %23 %6
         %25 = OpCompositeConstruct %v4float %8 %24 %float_0 %float_1
               OpStore %out_var_SV_Target0 %25
               OpReturn
               OpFunctionEnd
```

Removed the redundant OpCopyObjects and ended up with

```
#version 450

layout(location = 0) out vec4 out_var_SV_Target0;

void main()
{
    vec3 _22 = gl_FragCoord.xyz;
    mediump vec3 mp_copy_22 = _22;
    mediump vec3 _4 = mp_copy_22 * vec3(1.00986194610595703125);
    mediump float _5 = _4.x;
    float hp_copy_5 = _5;
    mediump float _6 = _4.y;
    float hp_copy_6 = _6;
    out_var_SV_Target0 = vec4((_5 * _6) * _6, (hp_copy_5 * hp_copy_6) * _6, 0.0, 1.0);
}
```

**Comment by Dredhog** [comment]
So far so good!

I tried it on a handful of shaders I had prepared during earlier testing, some of which are larger and more complex, went over them instruction by instruction (I even tested the `--force-temporary` option), also compared the Mali Offline Compiler perf reports  and everything looks correct in my eyes.

It will probably be a chunk of time before we can integrate this version into our engine for more exhaustive testing (currently only tried it via the CLI), but no need to wait for that, I'll make a separate report if I hit anything then.

Again, thank you very much for taking this on and great work!
